lib/mpl_toolkits/axisartist/grid_helper_curvelinear.py

```diff
@@ -16,29 +16,76 @@
+        differentiation step size is increased until at least one component of df/dx
+        is nonzero, under the constraint of not going out of the *xlims*, *ylims*
+        bounds.  If the gridline at a point is actually null (and the angle is thus not
+        well defined), the derivatives are evaluated after taking a small step along y;
+        this ensures e.g. that the tick at r=0 on a radial axis of a polar plot is
+        parallel with the ticks at r!=0.
+    thetas_dy
+        Like *thetas_dx*, but for df/dy.
     """
-    eps = np.finfo(float).eps ** (1/2)  # see e.g. scipy.optimize.approx_fprime
+
+    shape = np.broadcast_shapes(np.shape(xs), np.shape(ys))
     val = func(xs, ys)
-    # Take the finite difference step in the direction where the bound is the
-    # furthest; the step size is min of epsilon and distance to that bound.
-    xlo, xhi = sorted(xlims)
-    dxlo = xs - xlo
-    dxhi = xhi - xs
-    xeps = (np.take([-1, 1], dxhi >= dxlo)
-            * np.minimum(eps, np.maximum(dxlo, dxhi)))
-    val_dx = func(xs + xeps, ys)
-    ylo, yhi = sorted(ylims)
-    dylo = ys - ylo
-    dyhi = yhi - ys
-    yeps = (np.take([-1, 1], dyhi >= dylo)
-            * np.minimum(eps, np.maximum(dylo, dyhi)))
-    val_dy = func(xs, ys + yeps)
-    return (val, (val_dx - val) / xeps, (val_dy - val) / yeps)
+
+    thetas = []
+    eps0 = np.finfo(float).eps ** (1/2)  # cf. scipy.optimize.approx_fprime
+
+    # Take finite difference steps towards the furthest bound; the step size will be the
+    # min of epsilon and the distance to that bound.
+    xlo, xhi = sorted(xlim)
+    xdlo = xs - xlo
+    xdhi = xhi - xs
+    xeps_max = np.maximum(xdlo, xdhi)
+    xeps0 = np.where(xdhi >= xdlo, 1, -1) * np.minimum(eps0, xeps_max)
+    ylo, yhi = sorted(ylim)
+    ydlo = ys - ylo
+    ydhi = yhi - ys
+    yeps_max = np.maximum(ydlo, ydhi)
+    yeps0 = np.where(ydhi >= ydlo, 1, -1) * np.minimum(eps0, yeps_max)
+
+    for dfunc, ps, eps0, eps_max, eps_q in [
+            (lambda eps_p, eps_q: func(xs + eps_p, ys + eps_q),
+             xs, xeps0, xeps_max, yeps0),
+            (lambda eps_p, eps_q: func(xs + eps_q, ys + eps_p),
+             ys, yeps0, yeps_max, xeps0),
+    ]:
+        thetas_dp = np.full(shape, np.nan)
+        missing = np.full(shape, True)
+        eps_p = eps0
+        for it, eps_q in enumerate([0, eps_q]):
+            while missing.any() and (abs(eps_p) < eps_max).any():
+                if it == 0 and (eps_p > 1).any():
+                    break  # Degenerate derivative, move a bit along the other coord.
+                eps_p = np.minimum(eps_p, eps_max)
+                df_x, df_y = (dfunc(eps_p, eps_q) - dfunc(0, eps_q)) / eps_p
+                good = missing & ((df_x != 0) | (df_y != 0))
+                thetas_dp[good] = np.arctan2(df_y, df_x)[good]
+                missing &= ~good
+                eps_p *= 2
+        thetas.append(thetas_dp)
+    return (val, *thetas)
```

Comment:
Optional: It may be slightly more readable to to turn the outer loop into a function `calc_theta_d()` (name t.b.d.)
and rewrite this here to
```suggestion
    thetas_dx = calc_theta_d(
        lambda eps_p, eps_q: func(xs + eps_p, ys + eps_q),
        xeps0, xeps_max, yeps0, shape
    )
    thetas_dy = calc_theta_d(
        lambda eps_p, eps_q: func(xs + eps_q, ys + eps_p),
        yeps0, yeps_max, xeps0, shape
    )
    return val, (thetas_dx, thetas_dy)
```

with
```python
def calc_theta_d(dfunc, eps0, eps_max, eps_q, shape):
    thetas_dp = np.full(shape, np.nan)
    missing = np.full(shape, True)
    eps_p = eps0
    for it, eps_q in enumerate([0, eps_q]):
        while missing.any() and (abs(eps_p) < eps_max).any():
            if it == 0 and (eps_p > 1).any():
                break  # Degenerate derivative, move a bit along the other coord.
            eps_p = np.minimum(eps_p, eps_max)
            df_x, df_y = (dfunc(eps_p, eps_q) - dfunc(0, eps_q)) / eps_p
            good = missing & ((df_x != 0) | (df_y != 0))
            thetas_dp[good] = np.arctan2(df_y, df_x)[good]
            missing &= ~good
            eps_p *= 2
    return thetas_dp
```

Reply:
Agreed it looks nicer; changed.